Feature/kas 3360 aanpassen van vertrouwelijkheidstoggle op document niveau naar extra toegangsniveaus

sergiofenoll opened the pull request:
Frontend PR: https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1318
themis-export-service PR: https://github.com/kanselarij-vlaanderen/themis-export-service/pull/2
yggdrasil PR: https://github.com/kanselarij-vlaanderen/yggdrasil/pull/33

Remove confidentiality from pieces and add new access-levels

sergiofenoll committed:
Add migration for new access levels

Also makes it so confidential documents get Ministerraad access level
instead of being confidential

sergiofenoll committed:
Remove confidentiality & use :integer for priority

sergiofenoll committed:
Set right order of access levels

sergiofenoll committed:
Remove confidentiality on all graphs

erikap commented on config/migrations/20220427151516-new-access-levels-and-remove-confidentiality.sparql:
```diff
@@ -0,0 +1,67 @@
+  }
+};
+
+DELETE DATA {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/d335f7e3-aefd-4f93-81a2-1629c2edafa3> ext:prioriteit "1" . # Intern Regering
+    <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/abe4c18d-13a9-45f0-8cdd-c493eabbbe29> ext:prioriteit "2" . # Intern Overheid
+    <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/6ca49d86-d40f-46c9-bde3-a322aa7e5c8e> ext:prioriteit "3" . # Publiek
+  }
+};
+
+INSERT DATA {
+  GRAPH <http://mu.semte.ch/graphs/public> {
+    <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/d335f7e3-aefd-4f93-81a2-1629c2edafa3> ext:prioriteit 3 . # Intern Regering
+    <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/abe4c18d-13a9-45f0-8cdd-c493eabbbe29> ext:prioriteit 4 . # Intern Overheid
+    <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/6ca49d86-d40f-46c9-bde3-a322aa7e5c8e> ext:prioriteit 5 . # Publiek
+  }
+};
+
```
Codelists are currently stored in multiple graphs thus must be updated in multiple graphs as well. Although public graph is probably sufficient I would insert the new access-levels in both public- as kanselarij graph to be consistent with the past. Clearing out the required graphs for the codelists is a separate story (and if needed, redundant data will be removed in that story)

sergiofenoll replied:
These codelists aren't duplicated over multiple graphs. Previous migrations only insert them in the public graph, and querying locally I can't find them in the kanselarij graph. Do we still want to add these to the kanselarij graph, both the new as well as the old access levels?

erikap replied:
Sorry, didn't know you checked that already. No need to duplicate in kanselarij-graph in that case. And it's a good indicator that the other codelists can be removed from kanselarij-graph as well :crossed_fingers:

erikap commented on config/resources/dossier-domain.lisp:
```diff
@@ -116,7 +116,7 @@
                 (:scope-note  :string ,(s-prefix "skos:scopeNote"))
                 (:alt-label   :string ,(s-prefix "skos:altLabel"))
-                (:priority    :string ,(s-prefix "ext:prioriteit")))
+                (:priority    :integer ,(s-prefix "ext:prioriteit")))
```
Can we change to `schema:position` as well?  Impact for that change in this PR seems limited.

sergiofenoll committed:
Use schema:position for access levels

Still leaves the resource unchanged, the property is still called
priority. That change will come later

ValenberghsSven commented on config/migrations/20220427151516-new-access-levels-and-remove-confidentiality.sparql:
```diff
@@ -0,0 +1,68 @@
+}
+WHERE {
+  GRAPH ?g {
+    ?document a dossier:Stuk ;
+              ext:vertrouwelijk "true"^^<http://mu.semte.ch/vocabularies/typed-literals/boolean> ;
+              ext:toegangsniveauVoorDocumentVersie ?oldAccessLevel .
+  }
```
Having both `confidential: true` and `must have an existing accessLevel` will miss all pieces that don't have an accessLevel (mostly legacy before the default was set automatically).

queried on ACC:
current select: 5362 triples
with optional `?oldAccessLevel`: 21984 triples

sergiofenoll replied:
Thanks, didn't realize there were documents that lacked an access level.

ValenberghsSven reviewed (commented):
I'm a tad bit concerned about changing old data.

This line is vague: "het toegangsniveau van bestaande documenten die als vertrouwelijk gemarkeerd zijn wordt aangepast naar ‘Ministerraad’"
There is no consideration towards the current level.
fe: What if a document was confidential but had access level "intern overheid? In our current iteration of the app, that document is not visible with yggdrasil rules. But I recall it propably was visible once.

Out of the 5362 triples that are confidential and have atleast an accessLevel, 237 have an accessLevel that is not "intern regering". Can we just replace those with "Ministerraad"?

query to get that data:

```
PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX dossier: <https://data.vlaanderen.be/ns/dossier#>
PREFIX schema: <http://schema.org/>

select COUNT(*)WHERE {
  GRAPH ?g {
    ?document a dossier:Stuk ;
  ext:vertrouwelijk "true"^^<http://mu.semte.ch/vocabularies/typed-literals/boolean> .
    ?document ext:toegangsniveauVoorDocumentVersie ?oldAccessLevel .
    FILTER NOT EXISTS {?document ext:toegangsniveauVoorDocumentVersie <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/d335f7e3-aefd-4f93-81a2-1629c2edafa3> . }
  }
}
```
Changing "intern regering" to "ministerraad" with new yggdrasil also means Kabinet profile will lose reading rights to a bunch of documents, is that intended? Are those documents "leaks" in our current app?

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3360-aanpassen-van-vertrouwelijkheidstoggle-op-document-niveau-naar-extra-toegangsniveaus

sergiofenoll commented:
> I'm a tad bit concerned about changing old data.
> 
> This line is vague: "het toegangsniveau van bestaande documenten die als vertrouwelijk gemarkeerd zijn wordt aangepast naar ‘Ministerraad’" There is no consideration towards the current level. fe: What if a document was confidential but had access level "intern overheid? In our current iteration of the app, that document is not visible with yggdrasil rules. But I recall it propably was visible once.
> 
> Out of the 5362 triples that are confidential and have atleast an accessLevel, 237 have an accessLevel that is not "intern regering". Can we just replace those with "Ministerraad"?
> 
> query to get that data:
> 
> ```
> PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
> PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
> PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
> PREFIX dossier: <https://data.vlaanderen.be/ns/dossier#>
> PREFIX schema: <http://schema.org/>
> 
> select COUNT(*)WHERE {
>   GRAPH ?g {
>     ?document a dossier:Stuk ;
>   ext:vertrouwelijk "true"^^<http://mu.semte.ch/vocabularies/typed-literals/boolean> .
>     ?document ext:toegangsniveauVoorDocumentVersie ?oldAccessLevel .
>     FILTER NOT EXISTS {?document ext:toegangsniveauVoorDocumentVersie <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/d335f7e3-aefd-4f93-81a2-1629c2edafa3> . }
>   }
> }
> ```
> 
> Changing "intern regering" to "ministerraad" with new yggdrasil also means Kabinet profile will lose reading rights to a bunch of documents, is that intended? Are those documents "leaks" in our current app?

Maybe I am misunderstanding, but doesn't changing the underlying access level have no impact in practice? Yggdrasil takes both the confidentiality flag and the access level into account when determining which files to collect for later copying.

Going over how each type works right now, before merging the PR related to this ticket:
- For the [minister](https://github.com/kanselarij-vlaanderen/yggdrasil/blob/master/repository/distributors/minister-distributor.js#L102), currently nothing is checked, as they can access all files
- For the [cabinet](https://github.com/kanselarij-vlaanderen/yggdrasil/blob/master/repository/distributors/cabinet-distributor.js#L118), we check for confidentiality. The underlying access level is not taken into account.
- For the [government](https://github.com/kanselarij-vlaanderen/yggdrasil/blob/master/repository/distributors/government-distributor.js#L125), we check both that the access level isn't "Intern regering" and that the file is not marked as confidential. But this means that any file marked confidential, regardless of the underlying access level will not be copied.

In practice this means that once a file is confidential, the underlying access level does not really matter. For cabinet, if there were files before that had an access level "Intern regering" and confindential, they never had reading rights to begin with, so changing to "Ministerraad" access level should have no impact.

sergiofenoll committed:
Deal with legacy data that lacks access levels

ValenberghsSven commented:
> > I'm a tad bit concerned about changing old data.
> > This line is vague: "het toegangsniveau van bestaande documenten die als vertrouwelijk gemarkeerd zijn wordt aangepast naar ‘Ministerraad’" There is no consideration towards the current level. fe: What if a document was confidential but had access level "intern overheid? In our current iteration of the app, that document is not visible with yggdrasil rules. But I recall it propably was visible once.
> > Out of the 5362 triples that are confidential and have atleast an accessLevel, 237 have an accessLevel that is not "intern regering". Can we just replace those with "Ministerraad"?
> > query to get that data:
> > ```
> > PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
> > PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
> > PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
> > PREFIX dossier: <https://data.vlaanderen.be/ns/dossier#>
> > PREFIX schema: <http://schema.org/>
> > 
> > select COUNT(*)WHERE {
> >   GRAPH ?g {
> >     ?document a dossier:Stuk ;
> >   ext:vertrouwelijk "true"^^<http://mu.semte.ch/vocabularies/typed-literals/boolean> .
> >     ?document ext:toegangsniveauVoorDocumentVersie ?oldAccessLevel .
> >     FILTER NOT EXISTS {?document ext:toegangsniveauVoorDocumentVersie <http://kanselarij.vo.data.gift/id/concept/toegangs-niveaus/d335f7e3-aefd-4f93-81a2-1629c2edafa3> . }
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Changing "intern regering" to "ministerraad" with new yggdrasil also means Kabinet profile will lose reading rights to a bunch of documents, is that intended? Are those documents "leaks" in our current app?
> 
> Maybe I am misunderstanding, but doesn't changing the underlying access level have no impact in practice? Yggdrasil takes both the confidentiality flag and the access level into account when determining which files to collect for later copying.
> 
> Going over how each type works right now, before merging the PR related to this ticket:
> 
>     * For the [minister](https://github.com/kanselarij-vlaanderen/yggdrasil/blob/master/repository/distributors/minister-distributor.js#L102), currently nothing is checked, as they can access all files
> 
>     * For the [cabinet](https://github.com/kanselarij-vlaanderen/yggdrasil/blob/master/repository/distributors/cabinet-distributor.js#L118), we check for confidentiality. The underlying access level is not taken into account.
> 
>     * For the [government](https://github.com/kanselarij-vlaanderen/yggdrasil/blob/master/repository/distributors/government-distributor.js#L125), we check both that the access level isn't "Intern regering" and that the file is not marked as confidential. But this means that any file marked confidential, regardless of the underlying access level will not be copied.
> 
> 
> In practice this means that once a file is confidential, the underlying access level does not really matter. For cabinet, if there were files before that had an access level "Intern regering" and confindential, they never had reading rights to begin with, so changing to "Ministerraad" access level should have no impact.

hmmm, looking at it again with your explanation, I think you are right.
The subcase confidentiality is covered better/different in the new yggdrasil and works differently then I first thought.
 What I assumed:
 Subcase is confidential, but one of it's piece is made public > visible for all profiles.
 
 What is actually happens (I think, need to confirm to be 100% sure):
 Subcase is confidential so public pieces are not propagated of that subcase.

erikap commented:
> What actually happens (I think, need to confirm to be 100% sure): Subcase is confidential so "intern overheid" pieces are not propagated of that subcase.

That's a case I also discovered 2 weeks ago and started questioning. Johan verified with the users, and the confidentiality on a subcase should **not** have an impact on documents being propagated. Only the access level must be taken into account. But there is another story (https://kanselarij.atlassian.net/browse/KAS-3398) to change that behaviour in Yggdrasil.

erikap committed:
Bump Yggdrasil
